src/java.base/unix/classes/sun/nio/fs/UnixSecureDirectoryStream.java

```diff
@@ -414,15 +414,24 @@ public void setPermissions(Set<PosixFilePermission> perms)
                 if (!ds.isOpen())
                     throw new ClosedDirectoryStreamException();
 
-                int fd = (file == null) ? dfd : open();
-                try {
-                    fchmod(fd, UnixFileModeAttribute.toUnixMode(perms));
-                } catch (UnixException x) {
-                    x.rethrowAsIOException(file);
-                } finally {
-                    if (file != null && fd >= 0)
-                        UnixNativeDispatcher.close(fd, e-> null);
+                if (file == null)
+                    fchmod(dfd, UnixFileModeAttribute.toUnixMode(perms));
+                else if (followLinks)
+                    fchmodat(dfd, file, UnixFileModeAttribute.toUnixMode(perms), 0);
+                else if (fchmodatNoFollowSupported())
+                    fchmodat(dfd, file, UnixFileModeAttribute.toUnixMode(perms),
+                             AT_SYMLINK_NOFOLLOW);
+                else {
+                    int fd = open();
+                    try {
+                        fchmod(fd, UnixFileModeAttribute.toUnixMode(perms));
+                    } finally {
+                        if (fd >= 0)
+                            UnixNativeDispatcher.close(fd, e-> null);
+                    }
                 }
+            } catch (UnixException x) {
+                x.rethrowAsIOException(file);
```

Comment:
'file' will be null when changing the directory's permission, can you double check that case?